### PR TITLE
Add Admin RPC Front End for Top N Secondary Index Key Sizes Query.

### PR DESCRIPTION
**Pull request by IntokuSatori**
#### Problem

Need to expose the front end RPC API for https://github.com/solana-labs/solana/issues/28666

#### Summary of Changes

Add Admin RPC API.

### CODE REVIEW

**Comment by IntokuSatori** [comment]
Changed to a Draft PR till https://github.com/solana-labs/solana/pull/28887 goes in.

File: validator/src/admin_rpc_service.rs [hunk 1]
```diff
@@ -429,6 +443,45 @@ impl AdminRpc for AdminRpcImpl {
             Ok(found_sizes)
         })
     }
+
+    fn get_largest_index_keys(
+        &self,
+        meta: Self::Metadata,
+        secondary_index: RpcAccountIndex,
+        max_entries: usize,
+    ) -> Result<RpcResponse<Vec<(usize, String)>>> {
```
Comment (IntokuSatori) [inline validator/src/admin_rpc_service.rs:452]: Change to `Vec<(String, usize)>`

**Comment by IntokuSatori** [comment]
@CriesofCarrots Only file changed from https://github.com/solana-labs/solana/pull/28887 is `admin_rpc_service.rs`, This PR should update when I merge that one. Just waiting for the build to finish on the other PR. You can review that file tho now in case I need to make any changes. It returns the list as `Vec<(Pubkey, usize)>` as you mentioned in the other PR.

File: validator/src/admin_rpc_service.rs [hunk 2]
```diff
@@ -212,6 +216,16 @@ pub trait AdminRpc {
...
+
+    // This method does not itself do an accounts scan, but returns data only relevant to nodes
+    // that do support accounts-scan RPC apis
+    #[rpc(meta, name = "getLargestIndexKeys")]
+    fn get_largest_index_keys(
+        &self,
+        meta: Self::Metadata,
+        secondary_index: RpcAccountIndex,
+        max_entries: usize,
+    ) -> Result<RpcResponse<Vec<(String, usize)>>>;
```
Comment (CriesofCarrots) [inline validator/src/admin_rpc_service.rs:228]: I removed RpcResponse from `get_secondary_index_key_size` when I moved it to admin_rpc_service, my rationale being that these endpoints are to give a general picture of the secondary indexes and the exact slot in time is not super important. Wdyt? If in agreement, please make the return types consistent.

File: validator/src/admin_rpc_service.rs [hunk 3]
```diff
@@ -212,6 +216,16 @@ pub trait AdminRpc {
         meta: Self::Metadata,
         pubkey_str: String,
     ) -> Result<HashMap<RpcAccountIndex, usize>>;
+
+    // This method does not itself do an accounts scan, but returns data only relevant to nodes
+    // that do support accounts-scan RPC apis
```
Comment (CriesofCarrots) [inline validator/src/admin_rpc_service.rs:221]: Legacy comment?

File: validator/src/admin_rpc_service.rs (hunk 2 above)
Comment (CriesofCarrots) [reply validator/src/admin_rpc_service.rs:228]: Also, see recommendation below to convert Vec to HashMap

File: validator/src/admin_rpc_service.rs [hunk 4]
```diff
@@ -431,6 +445,40 @@ impl AdminRpc for AdminRpcImpl {
...
+                    context: RpcResponseContext::new(bank.slot()),
+                    value: Vec::new(),
+                });
+            };
+            let accounts_index = &bank.accounts().accounts_db.accounts_index;
+            let largest_keys = accounts_index
+                .get_largest_keys(&secondary_index, max_entries)
+                .iter()
+                .map(|&(x, y)| (y.to_string(), x))
+                .collect::<Vec<_>>();
```
Comment (CriesofCarrots) [inline validator/src/admin_rpc_service.rs:475]:
```suggestion
                .collect::<HashMap<_>>();
```

File: validator/src/admin_rpc_service.rs [hunk 5]
```diff
@@ -1012,4 +1071,183 @@ mod tests {
...
+
+        // Now add a random number of accounts each owned by one of the newely
+        // created dummy accounts
+        for dummy_account in &dummy_account_pubkeys {
+            // Add child accounts to each dummy account
+            let num_children = (&mut num_generator).sample_iter(key_size_range).next();
+            for j in 0..num_children.unwrap_or(0) {
+                let child_pubkey = Pubkey::new_unique();
+                let child_account = AccountSharedData::from(Account {
+                    lamports: ((j as u64) + 1) * 1000,
```
Comment (CriesofCarrots) [inline validator/src/admin_rpc_service.rs:1122]: I see that I missed this in your other PR, but these won't necessarily all be rent-exempt. I don't see that the lamports needs to vary for any particular reason, so I'd recommend just adding this somewhere above to use here and on L1107:
```
let lamports = bank.get_minimum_balance_for_rent_exemption(0);
```

File: validator/src/admin_rpc_service.rs [hunk 6]
```diff
@@ -1012,4 +1071,183 @@ mod tests {
...
+
+        // Make sure key lists conform to expected output
+        for key_list in largest_keys {
+            // No longer than the max
+            assert!(key_list.len() <= MAX_NUM_LARGEST_INDEX_KEYS_RETURNED);
+            let key_list_pubkeys = key_list
+                .iter()
+                .map(|(k, _)| Pubkey::from_str(k).unwrap())
+                .collect::<Vec<Pubkey>>();
+            // In sorted order: Decending key size, where ties are sorted by decening pubkey
```
Comment (CriesofCarrots) [inline validator/src/admin_rpc_service.rs:1244]: Spelling

File: validator/src/admin_rpc_service.rs (hunk 2 above)
Comment (IntokuSatori) [reply validator/src/admin_rpc_service.rs:228]: Not very opinionated on this. I tend to agree considering I don't know how the typical user would use this call besides on startup to filter accounts. In that context, exact slot doesn't seem too important. I must have missed the type removal on the other RPC call or I would have done the same here.

File: validator/src/admin_rpc_service.rs (hunk 3 above)
Comment (IntokuSatori) [reply validator/src/admin_rpc_service.rs:221]: Oh it was above the call for `get_secondary_index_key_size` too when I first made the PR, must have been removed at some point. I'll remove here as well.

File: validator/src/admin_rpc_service.rs (hunk 5 above)
Comment (IntokuSatori) [reply validator/src/admin_rpc_service.rs:1122]: It was for my own debugging initially. Gave me a differentiator on child accounts to verify the logic. Can probably remove now.

File: validator/src/admin_rpc_service.rs [hunk 7]
```diff
@@ -431,6 +445,40 @@ impl AdminRpc for AdminRpcImpl {
             Ok(found_sizes)
         })
     }
+
+    fn get_largest_index_keys(
+        &self,
+        meta: Self::Metadata,
+        secondary_index: RpcAccountIndex,
+        max_entries: usize,
+    ) -> Result<RpcResponse<Vec<(String, usize)>>> {
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply validator/src/admin_rpc_service.rs:454]: Decided in DM to leave this as Vec to preserve order

File: validator/src/admin_rpc_service.rs (hunk 4 above)
Comment (IntokuSatori) [reply validator/src/admin_rpc_service.rs:466]: https://github.com/solana-labs/solana/pull/29352#discussion_r1092634701

**Review by CriesofCarrots — APPROVED** [review]

**Comment by mvines** [comment]
Hey @IntokuSatori, your other two PRs made it into `master` before we cut the v1.15 branch but this PR just narrowly missed it. I suspect you'd prefer this feature to ship in the v1.15 release rather than waiting for v1.16, and if so just add the "v1.15" label to this PR to initiate a backport PR

**Comment by IntokuSatori** [comment]
> Hey @IntokuSatori, your other two PRs made it into `master` before we cut the v1.15 branch but this PR just narrowly missed it. I suspect you'd prefer this feature to ship in the v1.15 release rather than waiting for v1.16, and if so just add the "v1.15" label to this PR to initiate a backport PR

Backport complete! Should be good to go in v1.15 now.

**Pull request merged**